### ISSUE TITLE
Unable to update album info

### ISSUE THREAD

**Issue by NatureMuncher** [ampache/ampache#2141]
## Description
GUI appears to accept input but changes do not show up upon refresh.

### Describe the bug
Changing Album artist to correct value doesn't seem to stick
### To reproduce

1. Go to Album
2. Click on Edit
3. Scroll down to Artist
4. Click incorrect value
5. Scroll to and click on correct value
6. Click Save
7. Go back to album description and see old Artist value still exists

### Expected behavior
The value should update to the new one

## Screenshot
![image](https://user-images.githubusercontent.com/1648745/68505545-1954e780-022d-11ea-9c19-ddacc0e8c5ef.png)

## Environment
* Ampache version: Develop
* Web server + version: Apache/2.4.6
* Server operating system: CentOS Linux release 7.7.1908
* Client operating system: Fedora release 31 (ThirtyOne)

### Client type:
* Web

## Settings
https://pastebin.com/SvT80nzm

## Logs
```
2019-11-08 13:30:51 [ampache] (session.class) -> Existing session found 
2019-11-08 13:30:51 [mike] (session.class) -> c2145892c23b1f05bf448c283ff27140 has been extended to Fri, 08 Nov 2019 14:30:51 -0600 extension leng
th 3600 
2019-11-08 13:30:51 [mike] (edit.server) -> Called for action: {edit_object} 
2019-11-08 13:30:51 [mike] (tag.class) -> Updating tags for values {Other} type {album} object_id {860} 
2019-11-08 13:30:51 [ampache] (session.class) -> Existing session found 
2019-11-08 13:30:51 [mike] (session.class) -> c2145892c23b1f05bf448c283ff27140 has been extended to Fri, 08 Nov 2019 14:30:51 -0600 extension leng
th 3600 
2019-11-08 13:30:51 [mike] (session.class) -> Session created: 53ced1c92250377107327a5baa2de508 
2019-11-08 13:30:51 [mike] (browse.class) -> Show objects called for type {song} 
2019-11-08 13:30:51 [ampache] (session.class) -> Existing session found 
2019-11-08 13:30:51 [mike] (session.class) -> c2145892c23b1f05bf448c283ff27140 has been extended to Fri, 08 Nov 2019 14:30:51 -0600 extension leng
th 3600 
2019-11-08 13:30:51 [mike] (ajax.server) -> Called for page: {browse} 
2019-11-08 13:30:51 [mike] (browse.ajax) -> Called for action: {get_filters} 
2019-11-08 13:30:51 [mike] (query.class) -> set catalog id: 0 
```

**Comment by lachlan-00** [comment]
So you set the album artist and then close?
Then the album page doesn't change?

**Comment by NatureMuncher** [comment]
Correct. I select the artist from the drop-down, it looks correct. I click Save but when opening the album again the artist hasn't updated.
Is this another case where I need to perform a catalog update for the changes to be visible?

**Comment by lachlan-00** [comment]
No, that should make the change in the database. I will check this out

**Comment by Meinux** [comment]
Hi, 
same problem when trying to update an album. 
When deleting the Album number i have now 3 times the same album.

**Comment by Rhythm-of-Time** [comment]
Probably related: After editing an album (say id 18) and saving, I get the message: You have requested an Album that does not exist. URL reads https://ampache/#albums.php?action=show&album=18.

Actually the saved album now has **ID 19**. Edit it again, and it will become id 20... Can someone else reproduce and confirm this behaviour?

**Comment by lachlan-00** [comment]
Oh, what happens when you load the edit window. It preloaded the id of where to go so I need to redirect the changes.

**Comment by Rhythm-of-Time** [comment]
Did you see that the id gets altered in the album table, too? I wonder if it should be like this.

**Comment by lachlan-00** [comment]
this was just for album edit? 
was it from the album page?

I'll need to double check as i haven't found all of these possibilities. (I think i only fixed this for artist.)

**Comment by Rhythm-of-Time** [comment]
Yes, I tested from the album page. Just tried editing an artist. Works fine without altering ID.

**Comment by lachlan-00** [comment]
cool, i'm just trying to work out where i did this previously. that should help me test once i've found it.

**Comment by lachlan-00** [comment]
Just FYI for everyone here the ID is meant to change but has never redirected properly.

The sql id is used because it is unique but not for identification. i'm okay with how it currently works but need to work on the interface process

**Comment by Rhythm-of-Time** [comment]
There is an album ID displayed in Song Edit page:

![Song_Edit](https://user-images.githubusercontent.com/19483996/71055935-d35a4000-2157-11ea-9655-12f06c8db559.jpg)

That ID is not correct if you have made changes to the album via Album Edit. It is the ID the album had at the time the song was added. If you later change the artist, add tags or change the album title, each time you save the id counts up, see above. 

I had tried to get rid of all the orphans, but the **Song Editor** wouldn't let me. It does not save changes to  

The album ID 31 that gets displayed after clicking on More Metadata refers to the ID of my Unknown(Orphaned) album. In fact this is 32 in my case. I selected a previously created album from the dropdown list, but the album ID field remained unaffected. If I click save, the notorious loading div keeps going, but nothing else happens. On reloading the page everything is unchanged. 

The actual ID of the album selected from the list is 38. I tried again to set the name from the dropdown and entered 38 in the album ID field manually - no success. Also entering 32 (the current ID of the orphaned album) did not work. Surprisingly, the fields ArtistId and AlbumArtistId had been correctly set to the desired values (both ID 3), despite Artist and Album Fields being Unknown(Orphaned). So this seems to have been set correctly by previous saving.

I am quite confused about this.

**Comment by Rhythm-of-Time** [comment]
Finally, I found the culprit for this mess (sorry for the confused post above):

Firefox' console showed TypeError: resp.lastChild is undefined 
This is from tools.js, where this error had occured in Album Edit, but disappeared after fixing #2140

`resp:"<br />\n<b>Fatal error</b>: Uncaught ArgumentCountError: Too few arguments to function Song::update_artist(), 2 passed in /Ampache/lib/class/song.class.php on line 1172 and exactly 3 expected in /Ampache/lib/class/song.class.php:1444\nStack trace:\n#0 Ampache/lib/class/song.class.php(1172): Song::update_artist('3', '75')\n#1 /Ampache/server/edit.server.php(141)`

Guess what the missing argument is? It's the Album ID...

I'll see if there is a quick fix for this. @**lachlan-00** I think you are busy enough, but sure you'd be quicker. If you prefer to handle this yourself, please tell me.

**Comment by Rhythm-of-Time** [comment]
I thought the Album ID was missing, but song.class.php line 1444:

`public static function update_artist($new_artist, $song_id, $old_artist)`

reveals the argument $old_artist which is missing. 

And I see that new artists get generated and the stats get migrated. Seems a bit similar to album edit - is it also counting up the id? Anyway, this looks like a task for someone who understands Ampache code better than me...

**Comment by lachlan-00** [comment]
that where i come in!

good finds.

**Comment by lachlan-00** [comment]
can you see something in the log under song.class?
```
debug_event('song.class', $key . '=' . $value, 5);
```
old artist gets set to new artist.
$old_artist_id = $this->artist;

there isn't one?

**Comment by Rhythm-of-Time** [comment]
Those React-Tools for Firefox work incredibly well, but sometimes when debugging they slow down my machine in a way that I can't even type...

I only found these two:
2019-12-18 23:04:44 [admin] (edit.server) -> Called for action: {edit_object} 
2019-12-18 23:04:44 [admin] (song.class) -> title=01_Wetlands_MASTER 
2019-12-18 23:04:44 [admin] (song.class) -> artist=3

**Comment by lachlan-00** [comment]
artist=3
that's what i needed your original artist was 3, i'll keep looking

**Comment by Rhythm-of-Time** [comment]
And I checked from the database that the artist id has not changed.

**Comment by lachlan-00** [comment]
I see the answer and know what to fix.

it's artist being changed and it's calling a user_function, this argument doesn't have all the variables. i'll fix this today
```
                    // Check to see if it needs to be updated
                    if ($value != $this->$key) {
                        $function = 'update_' . $key;
                        self::$function($value, $this->id);
                        $this->$key = $value;
                    }
```

**Comment by Rhythm-of-Time** [comment]
Fine! Thanks.

**Comment by lachlan-00** [comment]
in develop, see what else you can find! ;)

**Comment by Rhythm-of-Time** [comment]
Wow. That **was** quick. Let me check.

**Comment by Rhythm-of-Time** [comment]
There is more in the logs, but the function is not working yet (changes are not saved).

2019-12-19 02:08:01 [admin] (edit.server) -> Called for action: {edit_object} 
2019-12-19 02:08:01 [admin] (song.class) -> title=22_RTA_Pass_MASTER 
2019-12-19 02:08:01 [admin] (song.class) -> artist=3 
2019-12-19 02:08:01 [admin] (song.class) -> album=38 
2019-12-19 02:08:01 [admin] (song.class) -> track=0 
2019-12-19 02:08:01 [admin] (song.class) -> mbid= 
2019-12-19 02:08:01 [admin] (song.class) -> composer= 
2019-12-19 02:08:01 [admin] (song.class) -> comment= 
2019-12-19 02:08:01 [admin] (song.class) -> label= 
2019-12-19 02:08:01 [admin] (song.class) -> year=0 
2019-12-19 02:08:01 [admin] (song.class) -> edit_tags= 
2019-12-19 02:08:01 [admin] (tag.class) -> Updating tags for values {} type {song} object_id {66} 
2019-12-19 02:08:01 [admin] (song.class) -> license=1 
2019-12-19 02:08:01 [admin] (song.class) -> metadata=Array 
2019-12-19 02:08:01 [admin] (song.class) -> id=66 
2019-12-19 02:08:01 [admin] (song.class) -> type=song_row

**Comment by lachlan-00** [comment]
which fields are you editing here? i'll do some edits myself

**Comment by Rhythm-of-Time** [comment]
I edit the artist name (select an existing artist from the dropdown) and the album in the same way. These changes are not saved. But what I can do now as opposed to before is edit the Album ID in the Metadata, this gets saved. I am just checking if I entered the right (current) id.

**Comment by Rhythm-of-Time** [comment]
Yes, it was set correctly to the current ID 38.

May this be an issue with the file location? My file organisation is a bit disordered after so many experiments yesterday. The file can be played, but the songs that I want to assign to the same album are in different locations. Don't know if this matters.

**Comment by lachlan-00** [comment]
what's interesting is that i created a new artist and it saved, then tried to set a new song to that artist and it failed.
I need to get home to check this out but can see there is an issue with changing the id instead of creating an id

**Comment by Rhythm-of-Time** [comment]
After reorganizing my files (created an upload folder, moved the existing user folders into it) and updating the catalogs I experienced something interesting:

The song I was previously trying to edit had now lost the information about existing artists and albums (artist 3 and album 38 were no more visible in the dropdown), so I had to create a new artist and album. This saved without problems, in Song View artist and album are now correctly displayed.

It seems you _must_ create an artist and an album first to be able to save these fields. The content of other fields can be changed and saved. Maybe this has to do with the album_name and artist_name cases (1144, 1151) in the switch? Artist::check($value) reads from where?

The comment `//Need to create new artist according the name` made me wonder why a new artist would need to be created if there is one in the list already. But if it _is_ in the dropdown list, you can't save this change.

**Comment by Rhythm-of-Time** [comment]
$new_artist_id = Artist::check($value); only works after Artist:: got updated by adding a new artist.

If not, it returns the cached value set in line 571 ff., which is in fact $old_artist_id. 

To get $new_artist_id in case of an existing artist, the value must be fetched from the dropdown instead. But function check_inline_song_edit(type, song) does not set the selected song. The new artist is displayed in the Artist field of Song Edit, but the original option value remains with selected="selected".

**Comment by wagnered** [comment]
I had a branch to work on problems with editing album data. But it disappeared before i had submitted a PR.  I remember discussing it in one of the issues.

In my mind, only the album artist can be changed.  I thought that  the artist had been included on the dialog because it was included in an SQL join query with the song table.  So the artist should not be on the album dialog.

Issue #564 brought the album editing problem back into my awareness..  I stumbled into this issue while looking for my original issue.

**Comment by lachlan-00** [comment]
should be very fixed now after all the major works here and with tests and regression checks on top
